<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/ros/test_ws/src/universal_robot/ur_description/urdf/ur5.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="ur5_robot" xmlns:xacro="http://ros.org/wiki/xacro">


    <xacro:property name="deg_to_rad" value="0.017453292519943295"/> <!-- pi/180 -->

    <!-- Include the table definition from another file -->
    <xacro:include filename="$(find pose_estimation)/urdf/table.urdf.xacro"/>

    <xacro:include filename="$(find pose_estimation)/urdf/end_effector.xacro"/>

    <transmission name="shoulder_pan_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="shoulder_pan_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="shoulder_pan_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <transmission name="shoulder_lift_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="shoulder_lift_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="shoulder_lift_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <transmission name="elbow_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="elbow_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="elbow_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <transmission name="wrist_1_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_1_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_1_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <transmission name="wrist_2_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_2_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_2_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <transmission name="wrist_3_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_3_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_3_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </actuator>
    </transmission>
    <!-- links: main serial chain -->
    <link name="base_link" />
    <link name="base_link_inertia">
        <visual>
            <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/base.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/base.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="4.0" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.00443333156" ixy="0.0" ixz="0.0" iyy="0.00443333156" iyz="0.0" izz="0.0072" />
        </inertial>
    </link>



    <!-- Fix the cobot to the world-->
    <link name="world"/>

    <!-- Fix the table to the world -->
    <joint name="world_table_fixed_joint" type="fixed">
        <parent link="world"/>
        <child link="table"/>
        <origin xyz="0 0 0.05" rpy="0 0 0"/> <!-- Adjust this to half the table thickness -->
    </joint>

    
    <!-- Now attach the robot to the table. -->
    <joint name="table_robot_fixed_joint" type="fixed">
        <parent link="table"/>
        <child link="base_link"/>
        <!-- The base_link of the robot is typically at the floor level,
             If we want it on the table, lift it by half the table thickness + any offset needed.
             For a table thickness of 0.05 m, let's try xyz="0 0 0.025" or just 0.05 if flush.
             Adjust as needed based on how the meshes align.
        -->
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <link name="shoulder_link">
        <visual>
            <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/shoulder.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/shoulder.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="3.7" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.014972352344389999" ixy="0.0" ixz="0.0" iyy="0.014972352344389999" iyz="0.0" izz="0.01040625" />
        </inertial>
    </link>
    <link name="upper_arm_link">
        <visual>
            <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.13585" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/upperarm.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.13585" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/upperarm.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="8.393" />
            <origin rpy="0 1.5707963267948966 0" xyz="-0.2125 0.0 0.136" />
            <inertia ixx="0.1338857818623325" ixy="0.0" ixz="0.0" iyy="0.1338857818623325" iyz="0.0" izz="0.0151074" />
        </inertial>
    </link>
    <link name="forearm_link">
        <visual>
            <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.0165" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/forearm.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.0165" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/forearm.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="2.275" />
            <origin rpy="0 1.5707963267948966 0" xyz="-0.196125 0.0 0.0165" />
            <inertia ixx="0.03121679102890359" ixy="0.0" ixz="0.0" iyy="0.03121679102890359" iyz="0.0" izz="0.004095" />
        </inertial>
    </link>
    <link name="wrist_1_link">
        <visual>
            <!-- TODO: Move this to a parameter -->
            <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.093" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/wrist1.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.093" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/wrist1.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="1.219" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.0020138887777775" ixy="0.0" ixz="0.0" iyy="0.0020138887777775" iyz="0.0" izz="0.0021942" />
        </inertial>
    </link>
    <link name="wrist_2_link">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 -0.095" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/wrist2.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 -0.095" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/wrist2.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="1.219" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.0018310388509175" ixy="0.0" ixz="0.0" iyy="0.0018310388509175" iyz="0.0" izz="0.0021942" />
        </inertial>
    </link>
    <link name="wrist_3_link">
        <visual>
            <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.0818" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/visual/wrist3.dae" />
            </geometry>
            <material name="LightGrey">
                <color rgba="0.7 0.7 0.7 1.0" />
            </material>
        </visual>
        <collision>
            <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.0818" />
            <geometry>
                <mesh filename="package://ur_description/meshes/ur5/collision/wrist3.stl" />
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1879" />
            <origin rpy="0 0 0" xyz="0.0 0.0 -0.01525" />
            <inertia ixx="8.062472608343e-05" ixy="0.0" ixz="0.0" iyy="8.062472608343e-05" iyz="0.0" izz="0.0001321171875" />
        </inertial>
    </link>
    <!-- joints: main serial chain -->
    <joint name="base_link-base_link_inertia" type="fixed">
        <parent link="base_link" />
        <child link="base_link_inertia" />
        <!-- 'base_link' is REP-103 aligned (so X+ forward), while the internal
           frames of the robot/controller have X+ pointing backwards.
           Use the joint between 'base_link' and 'base_link_inertia' (a dummy
           link/frame) to introduce the necessary rotation over Z (of pi rad).
      -->
        <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
    </joint>
    <joint name="shoulder_pan_joint" type="revolute">
        <parent link="base_link_inertia" />
        <child link="shoulder_link" />
        <origin rpy="0 0 0" xyz="0 0 0.089159" />
        <axis xyz="0 0 1" />
        <limit effort="150.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <joint name="shoulder_lift_joint" type="revolute">
        <parent link="shoulder_link" />
        <child link="upper_arm_link" />
        <origin rpy="1.570796327 0 0" xyz="0 0 0" />
        <axis xyz="0 0 1" />
        <limit effort="150.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <joint name="elbow_joint" type="revolute">
        <parent link="upper_arm_link" />
        <child link="forearm_link" />
        <origin rpy="0 0 0" xyz="-0.425 0 0" />
        <axis xyz="0 0 1" />
        <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <joint name="wrist_1_joint" type="revolute">
        <parent link="forearm_link" />
        <child link="wrist_1_link" />
        <origin rpy="0 0 0" xyz="-0.39225 0 0.10915" />
        <axis xyz="0 0 1" />
        <limit effort="28.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <joint name="wrist_2_joint" type="revolute">
        <parent link="wrist_1_link" />
        <child link="wrist_2_link" />
        <origin rpy="1.570796327 0 0" xyz="0 -0.09465 -1.941303950897609e-11" />
        <axis xyz="0 0 1" />
        <limit effort="28.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <joint name="wrist_3_joint" type="revolute">
        <parent link="wrist_2_link" />
        <child link="wrist_3_link" />
        <origin rpy="1.570796326589793 3.141592653589793 3.141592653589793" xyz="0 0.0823 -1.688001216681175e-11" />
        <axis xyz="0 0 1" />
        <limit effort="28.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793" />
        <dynamics damping="0" friction="0" />
    </joint>
    <!-- ROS-Industrial 'base' frame: base_link to UR 'Base' Coordinates transform -->
    <link name="base" />
    <joint name="base_link-base_fixed_joint" type="fixed">
        <!-- Note the rotation over Z of pi radians: as base_link is REP-103
           aligned (ie: has X+ forward, Y+ left and Z+ up), this is needed
           to correctly align 'base' with the 'Base' coordinate system of
           the UR controller.
      -->
        <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
        <parent link="base_link" />
        <child link="base" />
    </joint>
    <!-- ROS-Industrial 'flange' frame: attachment point for EEF models -->
    <link name="flange" />
    <joint name="wrist_3-flange" type="fixed">
        <parent link="wrist_3_link" />
        <child link="flange" />
        <origin rpy="0 -1.5707963267948966 -1.5707963267948966" xyz="0 0 0" />
    </joint>
    <!-- ROS-Industrial 'tool0' frame: all-zeros tool frame -->
    <link name="tool0" />
    <joint name="flange-tool0" type="fixed">
        <!-- default toolframe: X+ left, Y+ up, Z+ front -->
        <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0 0 0" />
        <parent link="flange" />
        <child link="tool0" />
    </joint>


    <!-- Attach a camera sensor to tool0 link -->
    <gazebo reference="tool0">
    <sensor type="camera" name="wrist_camera">
        <always_on>true</always_on>
        <update_rate>30.0</update_rate>
        <visualize>true</visualize>
        <camera name="head_camera">
        <horizontal_fov>1.047</horizontal_fov>
        <image>
            <width>640</width>
            <height>480</height>
            <format>R8G8B8</format>
        </image>
        <clip>
            <near>0.05</near>
            <far>10.0</far>
        </clip>
        </camera>
        <plugin name="gazebo_ros_camera" filename="libgazebo_ros_camera.so">
        <always_on>true</always_on>
        <update_rate>30.0</update_rate>
        <camera_name>head_camera</camera_name>
        <image_topic_name>image_raw</image_topic_name>
        <camera_info_topic_name>camera_info</camera_info_topic_name>
        <frame_name>tool0</frame_name>
        <hack_baseline>0.0</hack_baseline>
        <distortion_k1>0.0</distortion_k1>
        <distortion_k2>0.0</distortion_k2>
        <distortion_k3>0.0</distortion_k3>
        <distortion_t1>0.0</distortion_t1>
        <distortion_t2>0.0</distortion_t2>
        </plugin>
    </sensor>
    </gazebo>


    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/ur5</robotNamespace>
        </plugin>
    </gazebo>

</robot>